### PR TITLE
Finally a new command to request external books!

### PR DESCRIPTION
**Pull request by ellegint**
As title says, this new update let people make requests for adding books that aren't present in the UNICT catalogue.
This is made using an admin group where devs can accept or decline those requests with a button. The user is been updated about the admin's choice, either if the request has been accepted or declined.
I also modified the attributes name of the database's table making them in english, added some function to print books info and fixed some issue with ISBN found online. Infact, sometimes the scraper got too much information from the website. This update should fix that problem, giving us only a 13 digits ISBN.

### CODE REVIEW

File: module/button.py [hunk 1]
```diff
@@ -1,18 +1,67 @@
...
+        if rows:
+
+            chat_id = rows[0][1]
+
+            if vote == 'Y':
+                isbn = rows[0][2]
+                title = rows[0][3]
+                authors = rows[0][4]
+                username = rows[0][5]
+                price = rows[0][6]
```
Comment (Helias) [inline module/button.py]:
```suggestion
_, __, isbn, title, authors, username, price = rows[0]
```

Comment (Helias) [reply module/button.py]: would be better to set chat_id here and make the `chat_id = rows[0][1]` only inside `if vote == 'N'`

File: module/button.py [hunk 2]
```diff
@@ -1,18 +1,67 @@
...
+                query.edit_message_text(text=f"Richiesta accettata.")
+                add_book(isbn, title, authors)
+                add_item(isbn, title, authors, username, price)
+                delete_request(context, chat_id, row_id)
+                update_similar_requests(context, isbn)
+                context.bot.send_message(chat_id, "Il libro è stato messo in vendita.")
+            
+            if vote == 'N':
```
Comment (Helias) [inline module/button.py]: should be `elif vote == 'N'`?

File: module/button.py [hunk 3]
```diff
@@ -1,18 +1,67 @@
...
+                query.edit_message_text(text=f"Richiesta rifiutata.")
+                delete_request(context, chat_id, row_id)
+                context.bot.send_message(chat_id, "La tua richiesta è stata rifiutata. Controlla se i dati inseriti sono corretti e riprova.")
+        
+        else:
+            if vote == 'Y' or vote == 'N':
```
Comment (Helias) [inline module/button.py]: `else` and `if` can be converted into `elif`, but if you want to keep this for readability, leave as it is

File: module/button.py (hunk 1 above)
Comment (ellegint) [reply module/button.py]: So you suggest to write `_, chat_id, isbn, title, authors, username, price = rows[0]` in the `vote = Y` and put the `chat_id = rows[0][1]` in the `vote = N`?
By the way, chat_id is needed in both the if clauses, i thought it was a good idea to put it there

File: module/button.py (hunk 2 above)
Comment (ellegint) [reply module/button.py]: A vote can never be both Y and N simultaneously, so it works anyway. I can change it anyways if you want

File: module/button.py (hunk 3 above)
Comment (ellegint) [reply module/button.py]: Yeah, I think I'll keep it that way for readability

File: module/button.py [hunk 4]
```diff
@@ -1,18 +1,67 @@
...
+            query.edit_message_text(text=f"Eliminazione del libro selezionato...")
+            conn.commit()
+            context.bot.send_message(chat_id, "Libro eliminato.")
+        else:
+            query.edit_message_text(text=f"Si è verificato un problema nella lettura del database.")
+        conn.close()
+    
+    if query.data.split(';')[0] == 'new_request':
+        vote = query.data.split(';')[1]
+        row_id = query.data.split(';')[2]
```
Comment (Helias) [inline module/button.py]: avoid splitting multiple times the same data using the destructuring/unpack feature
```suggestion
        _, vote, row_id = query.data.split(';')
```

File: module/button.py [hunk 5]
```diff
@@ -1,18 +1,61 @@
...
+from module.manage_requests import delete_request, update_similar_requests
 
 
 def button(update: Update, context: CallbackContext) -> None:
     chat_id = update.effective_chat.id
     query = update.callback_query
-    conn = create_connection(DB_PATH)
-    if conn:
+    
+    if query.data.split(';')[0] == 'delete':
```
Comment (Helias) [inline module/button.py]: you have two identical if, they should be `if` and `elif`, moreover avoid to do the same big code `query.data.split(';')[0]`. save it into a variable or use `match` (it's `switch-case` in other languages)

Moreover, you could avoid magic strings like `delete` or `new_request`, you could save them into constant variables.

File: module/button.py [hunk 6]
```diff
@@ -1,18 +1,61 @@
...
-        conn.commit()
-        context.bot.send_message(chat_id, "Libro eliminato.")
-    else:
-        query.edit_message_text(text=f"Si è verificato un problema nella lettura del database.")
+        cur.execute("SELECT rowid, * FROM Requests WHERE rowid=?", (row_id,))
+        rows = cur.fetchall()
+        
+        if rows:
+
+            if vote == 'Y':
```
Comment (Helias) [inline module/button.py]: here and below, avoid magic strings like `'Y'` or `'N'`, save them into constant variable like `NO = 'N'` and `YES = 'Y'`

File: module/delete.py [hunk 7]
```diff
@@ -12,6 +12,6 @@ def delete(update: Update, context: CallbackContext) -> None:
     
     rows = get_user_books(context, chat_id)
     if rows:
-        keyboard = [[InlineKeyboardButton(str(i + 1), callback_data=(rows[i][0]))] for i in range(len(rows))]
+        keyboard = [[InlineKeyboardButton(str(i + 1), callback_data=('delete;'+str(rows[i][0])))] for i in range(len(rows))]
```
Comment (Helias) [inline module/delete.py]: this seems a big hack, but it's ok

File: module/request.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+        cur = conn.cursor()
+        cur.execute("SELECT * FROM Requests WHERE ISBN=? AND Seller=?", (user_isbn, username,))
+        rows = cur.fetchall()
+        
+        if not rows:
+            row_id = add_request(context, chat_id, user_isbn, title, authors, username, price)
+            send_request(context, row_id)
+            context.bot.send_message(chat_id, "La richiesta è stata inoltrata agli admin. Grazie del supporto!")
+        else:
+            context.bot.send_message(chat_id, "Hai già inviato una richiesta per questo libro. La tua richiesta è in elaborazione.")
```
Comment (Helias) [inline module/request.py]: the send message will be called in if and else, what changes is the text inside, let's create a variable for it
```suggestion
        if not rows:
            row_id = add_request(context, chat_id, user_isbn, title, authors, username, price)
            send_request(context, row_id)
            message_text = "La richiesta è stata inoltrata agli admin. Grazie del supporto!"
        else:
            message_text = "Hai già inviato una richiesta per questo libro. La tua richiesta è in elaborazione."

         context.bot.send_message(chat_id, message_text)
```

File: module/request.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+            title = soup.find("strong").text.split("/")[0]
+            authors = soup.find("strong").text.split("/")[1]
+            context.bot.send_message(chat_id, 'Il libro esiste già nel database locale:\n' + get_book_info(isbn, title, authors))
+            add_book(isbn, title, authors)
+            add_item(isbn, title, authors, username, price)
+            context.bot.send_message(chat_id, "Il libro è stato messo in vendita.")
+            return
+
+        title = message.split('; ')[2]
+        authors = message.split('; ')[3]
```
Comment (Helias) [inline module/request.py]:
```suggestion
        _, _, title, authors = message.split('; ')
```

File: module/sell.py [hunk 10]
```diff
@@ -1,20 +1,33 @@
...
+    if len(isbn) > 13:
+        p1 = re.search('978', isbn)
+        p2 = re.search('979', isbn)
+        if p1 is not None:
+            s = p1.span()[0]
+            isbn = isbn[s:s+13]
+        elif p2 is not None:
+            s = p2.span()[0]
+            isbn = isbn[s:s+13]
+    return isbn
```
Comment (Helias) [inline module/sell.py]: you like to use `if` and `elif` where it is not efficient to use 😁 
```suggestion
    isbn = isbn.replace('-','')
    if len(isbn) > 13:
        p1 = re.search('978', isbn)
        if p1 is not None:
            s = p1.span()[0]
            return isbn[s:s+13]
        
        p2 = re.search('979', isbn)
        if p2 is not None:
            s = p2.span()[0]
            return isbn[s:s+13]

    return isbn
```

File: module/manage_requests.py [hunk 11]
```diff
@@ -0,0 +1,75 @@
+import yaml
+from telegram import InlineKeyboardButton, InlineKeyboardMarkup
+from telegram.ext import CallbackContext
+from module.create_connection import create_connection
+from module.add_item import add_item
+from module.find import find
+from module.shared import DB_PATH, error_message
+
+
+def add_request(context: CallbackContext, chat_id: int, isbn: str, title: str, authors: str, username: str, price: str) -> None:
```
Comment (Helias) [inline module/manage_requests.py]: here you are not reeturning `None`

File: module/manage_requests.py [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+    sql = """ INSERT INTO Requests(ChatID, ISBN, Title, Authors, Seller, Price)
+              VALUES(?,?,?,?,?,?) """
+    conn = create_connection(DB_PATH)
+    if not conn:
+        context.bot.send_message(chat_id, error_message)
+        return
+    cur = conn.cursor()
+    cur.execute(sql, item)
+    conn.commit()
+    conn.close()
```
Comment (Helias) [inline module/manage_requests.py]: this code is really similar to the code below under delete_request, you could wrap it into a beautiful function and reduce the redundancy of the code
this is up to you, you could make a function that manages connection, execution, commit and closing of the conn, or keep it as it is

May you could do this in another PR if you prefer

File: module/button.py (hunk 5 above)
Comment (ellegint) [reply module/button.py]: Like the comment above, `query.data.split(';')[0]` will have only two values: `delete` or `new_request`. Only one of those ifs will compute. Hence, there's no need to put the elif.
Good idea for the constants, I'll add those to the `shared.py` module

File: module/manage_requests.py [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+    if not conn:
+        context.bot.send_message(group_id, error_message)
+        return
+    cur = conn.cursor()
+    cur.execute("SELECT ISBN, Title, Authors, Seller, Price FROM Requests WHERE rowid=?", (row_id,))
+    rows = cur.fetchall()
+    conn.close()
+    
+    isbn, title, authors, username, price = rows[0]
+    context.bot.send_message(group_id, "New Pending Request:\n"+isbn+"\n"+title+"\n"+authors+"\n"+username+"\n"+price+"\n", reply_markup=reply_markup)
```
Comment (Helias) [inline module/manage_requests.py]: use `join` to make it more elegant, ref https://www.w3schools.com/python/ref_string_join.asp
```suggestion
    isbn, title, authors, username, price = rows[0]
    request = 'New Pending Request: \n' + [isbn, title, authors, username, price].join('\n') + '\n'
    context.bot.send_message(group_id, request, reply_markup=reply_markup)
```

File: module/manage_requests.py [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+    cur = conn.cursor()
+    cur.execute("DELETE FROM Requests WHERE rowid=?", (row_id,))
+    conn.commit()
+    conn.close()
+
+def send_request(context: CallbackContext, row_id: int) -> None:
+    with open('config/settings.yaml', 'r') as yaml_config:
+            config_map = yaml.load(yaml_config, Loader=yaml.SafeLoader)
+    group_id = config_map['admin_group_id']
+    keyboard = [[InlineKeyboardButton('Y', callback_data=('new_request;Y;' + str(row_id)))], [InlineKeyboardButton('N', callback_data=('new_request;N;'+ str(row_id)))]]
```
Comment (Helias) [inline module/manage_requests.py]: this `'Y'` and `'N'` need to be in constants variables shared in the project ref https://testdriven.io/blog/clean-code-python/#6-dont-use-magic-numbers

File: module/request.py [hunk 15]
```diff
@@ -0,0 +1,88 @@
...
+            context.bot.send_message(chat_id, 'Il libro esiste già nel database locale:\n' + get_book_info(isbn, title, authors))
+            add_item(isbn, title, authors, username, price)
+            context.bot.send_message(chat_id, "Il libro è stato messo in vendita.")
+            return
+
+        found, soup = book_in_unict(user_isbn)
+        if found:
+            isbn = _get_isbn_from_website(soup)
+            title = soup.find("strong").text.split("/")[0]
+            authors = soup.find("strong").text.split("/")[1]
```
Comment (Helias) [inline module/request.py]: replace it using unpack feature

File: module/sell.py [hunk 16]
```diff
@@ -45,30 +58,27 @@ def sell(update: Update, context: CallbackContext) -> None:
...
             add_item(isbn, title, authors, username, price)
-            context.bot.send_message(chat_id, "Il libro è stato aggiunto al database.")
+            context.bot.send_message(chat_id, "Il libro è stato messo in vendita.")
             return
 
         found, soup = book_in_unict(user_isbn)
         if found:
             isbn = _get_isbn_from_website(soup)
             title = soup.find("strong").text.split("/")[0]
             authors = soup.find("strong").text.split("/")[1]
```
Comment (Helias) [inline module/sell.py]: unpack

File: module/send_results.py [hunk 17]
```diff
@@ -5,9 +5,17 @@
...
-        res += ("n°: " + str(i + 1) + "\n" + "ISBN: " + rows[i][1] + "\n" + "Titolo: " + rows[i][2] + "\n" + "Autori: " + rows[i][3] + "\n" + "Venditore: " + rows[i][4] + "\n" + "Prezzo: " + rows[i][5] + " €\n\n")
+        res += ("n°: " + str(i + 1) + "\n" + get_item_info(rows[i][1], rows[i][2], rows[i][3], rows[i][4], rows[i][5]) + "\n")
         if (i + 1) % 3 == 0:
             context.bot.send_message(chat_id, res)
             res = ""
     if res != "":
         context.bot.send_message(chat_id, res)
+
+def get_book_info(isbn: str, title: str, authors: str) -> str:
+    s = "ISBN: " + isbn + "\n" + "Titolo: " + title + "\n" + "Autori: " + authors + "\n"
```
Comment (Helias) [inline module/send_results.py]: you can use the join or you can make it simpler using `f-string`
```suggestion
    s = f"ISBN: {isbn} \nTitolo: {title}\nAutori: {authors}\n"
```

File: module/send_results.py [hunk 18]
```diff
@@ -5,9 +5,17 @@
...
+    return s
+
+def get_item_info(isbn: str, title: str, authors: str, seller: str, price: str) -> str:
+    s = "ISBN: " + isbn + "\n" + "Titolo: " + title + "\n" + "Autori: " + authors + "\n" + "Venditore: " + seller + "\n" + "Prezzo: " + price + " €\n"
```
Comment (Helias) [inline module/send_results.py]: same here, your choice to use `f-string` or `join`, probably f-string is more efficient

Comment (Helias) [reply module/send_results.py]: moreover, you could return the text directly without creating the `s` variable, like..

`return f"ISBN: {isbn}\nTitolo: {title}....etc`

File: module/button.py (hunk 5 above)
Comment (Helias) [reply module/button.py]: I don't know if I am missing something but there is no return in the first if, so the second if will be triggered anyway

So the second if will trigger again a split in `query.data.split(';')[0]` and a comparison with `'delete'`, with the elif you'll skip it, but anyway would be better to wrap query.data.split into a separate variable or into `match` and use constants

Comment (Helias) [reply module/button.py]:
```python
if something == 'asd':
   return
  
if something == 'lol':
  ...
```

that's ok

```python
if something == 'asd':
   ...
   
if something == 'lol':
  ...
```

this requires `elif`, but still, it's better `match` (switch case)

File: main.py [hunk 19]
```diff
@@ -3,10 +3,11 @@
 import logging, yaml
 from module.handlers import handlers
 from app import app
+from module.shared import YAML_PATH
 
 
 def main() -> None:
-    with open('config/settings.yaml', 'r') as yaml_config:
+    with open(YAML_PATH, 'r') as yaml_config:
```
Comment (Helias) [inline main.py]: nice one 👍

File: module/button.py [hunk 20]
```diff
@@ -2,60 +2,45 @@
 from telegram.ext import CallbackContext
 from module.add_book import add_book
 from module.add_item import add_item
-from module.create_connection import create_connection
-from module.shared import DB_PATH, error_message
-from module.add_book import add_book
-from module.add_item import add_item
+from module.create_connection import connect_and_execute
+from module.shared import *
```
Comment (Helias) [inline module/button.py]: not sure about using `*`

File: module/create_connection.py [hunk 21]
```diff
@@ -9,3 +11,29 @@ def create_connection(db_file: str) -> sqlite3.Connection:
     except Error as e:
         print(str(e))
     return conn
+
+
+def connect_and_execute(context: CallbackContext, chat_id: int, query: str, params: tuple, operation: str) -> Union[int, list, None]:
```
Comment (Helias) [inline module/create_connection.py]: I don't remember if python3.9 supports this
```suggestion
def connect_and_execute(context: CallbackContext, chat_id: int, query: str, params: tuple, operation: str) -> Optional[int | list]:
```

File: module/create_connection.py [hunk 22]
```diff
@@ -9,3 +11,29 @@ def create_connection(db_file: str) -> sqlite3.Connection:
...
+    
+    elif operation == DELETE:
+        conn.commit()
+        conn.close()
+        return
+
+    elif operation == SELECT:
+        rows = cur.fetchall()
+        conn.close()
+        return rows
```
Comment (Helias) [inline module/create_connection.py:39]: the `elif` can become `if`
if you use `return` in all `ifs` you can convert `elif` into `if`

Comment (Helias) [reply module/create_connection.py:39]: moreover, here you should use match for `operation` https://learnpython.com/blog/python-match-case-statement/

File: module/find.py [hunk 23]
```diff
@@ -1,8 +1,23 @@
...
+from module.shared import BOOKS, SELECT
 
 
-def find(txt: str, conn: sqlite3.Connection, s: str) -> List[tuple]:
+def find(context: CallbackContext, chat_id: int, txt: str, mode: str) -> Union[int, list, None]:
+    if mode == BOOKS:
+        query = "SELECT * FROM Books WHERE ISBN=?"
+        params = (txt,)
+    else:
+        q = "%" + txt + "%"
```
Comment (Helias) [inline module/find.py]:
```suggestion
        q = f"%{txt}%"
```

File: module/my_books.py [hunk 24]
```diff
@@ -1,35 +1,33 @@
...
+        context.bot.send_message(chat_id, NO_BOOKS)
     return rows
 
 
 def my_books(update: Update, context: CallbackContext) -> None:
     chat_id = update.effective_chat.id
     message = update.message.text
     if message != "/libri":
-        context.bot.send_message(chat_id, "Utilizzo comando: /libri")
+        context.bot.send_message(chat_id, MY_BOOKS_USAGE)
```
Comment (Helias) [inline module/my_books.py:32]: it was not necessary, but it's not an issue

File: module/shared.py [hunk 25]
```diff
@@ -1,5 +1,109 @@
...
+
+BOOK_NOT_AVAILABLE = "Libro non trovato. Controlla di aver inserito correttamente l'ISBN. Se l'ISBN è corretto, utilizza il comando /richiedi per fare una richiesta di inserimento manuale."
+
+
+#Scraping
+URL_1 = "https://catalogo.unict.it/search/i?SEARCH="
+
+URL_2 = "&sortdropdown=-&searchscope=9"
+
+NO_MATCHES = "No matches found"
```
Comment (Helias) [inline module/shared.py:109]: wonderful 🚀

File: module/button.py (hunk 20 above)
Comment (Helias) [reply module/button.py]: i fixed it

**Review by Helias — APPROVED** [review]
